### PR TITLE
AP_BattMonitor: SMBus: Fetch serial numbers and pack capacity

### PR DESCRIPTION
**Pull request by WickedShell**
This is another round of SMBus enhancements. The first one is that we now read the battery serial number, and stash it in a parameter to notify the user. (I didn't see any other convenient way to transport and log the value, and it's the same behavior we already do for capacity). We can't set_and_notify on a parameter from an timer context so the value is first stashed in the backend before being sent out.

The second commit is simply a cleanup I noticed while making the first change set. We kept a direct reference to the instance of the driver in the backend, but also stored the reference within the battery state object. It didn't make sense to have both references, I flipped a coin and the _state.instance got to live (although I could be convinced that having the instance reference live outside the state block might make more sense) :)

Finally I moved the fetching of full charge capacity into a common helper and leverage it within the Maxell driver as well. This is also an attempt to correct an outstanding bug on ArduPilot master with Solo batteries where we don't actually fetch the full charge capacity correctly. (Related to the fact that the default parameter value meant we never attempted to update the parameter).

The initial serial number pass was tested by @Pedals2Paddles, but some of the fetching of it was reworked, as well as the patches to correct full charge capacity. Those both need to be tested on a Solo before this can be merged. Whoever I can coerce into testing I'd really like to see a DF log as well that is done with LOG_DISARMED 1.

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 1]
```diff
@@ -2,7 +2,27 @@
...
+    _mon._serial_numbers[_state.instance] = AP_BATT_SERIAL_NUMBER_DEFAULT;
+}
+
+/// read the battery_voltage and current, should be called at 10hz
+void AP_BattMonitor_SMBus::read(void)
+{
+    // nothing to be done here for actually interacting with the battery
+    // however we can use this to set any parameters that need to be set
+
+    if (_serial_number && _mon._serial_numbers[_state.instance] == -1) {
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: Nitpicking: comparing *_serial_number* to 0 is nicer (since it isn't a boolean) and changing the order would probably lead to a faster evaluation of the condition on most of the times.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 2]
```diff
@@ -18,6 +38,22 @@ bool AP_BattMonitor_SMBus::read_temp(void)
     return false;
 }
 
+// reads the serial number if it's not already known
+// returns true if the read was successful or the number was already known
+bool AP_BattMonitor_SMBus::read_serial_number(void) {
+    uint16_t data;
+
+    // don't recheck the serial number if we already have it
+    if (_serial_number) {
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: Nitpicking: same about comparing to 0

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h [hunk 3]
```diff
@@ -18,21 +18,24 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
...
-    {}
+                    AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev);
 
     // virtual destructor to reduce compiler warnings
     virtual ~AP_BattMonitor_SMBus() {}
 
 
 protected:
 
+    void read(void);
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h]: `override` please.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 4]
```diff
@@ -2,15 +2,18 @@
...
+#define BATTMONITOR_SMBUS_SERIAL               0x1C // serial number
+
 AP_BattMonitor_SMBus::AP_BattMonitor_SMBus(AP_BattMonitor &mon,
                                            AP_BattMonitor::BattMonitor_State &mon_state,
                                            AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
         : AP_BattMonitor_Backend(mon, mon_state),
         _dev(std::move(dev))
 {
     _mon._serial_numbers[_state.instance] = AP_BATT_SERIAL_NUMBER_DEFAULT;
+    _mon._pack_capacity[_state.instance] = 0;
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: Why is this? Those are the default values, right?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 5]
```diff
@@ -22,6 +25,25 @@ void AP_BattMonitor_SMBus::read(void)
     if (_serial_number && _mon._serial_numbers[_state.instance] == -1) {
         _mon._serial_numbers[_state.instance].set_and_notify(_serial_number);
     }
+
+    if (_full_charge_capacity != _mon._pack_capacity[_state.instance]) {
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: I'd prefer it did the same check as above: check if the parameter has the default value and whether we got the full charge capacity.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 6]
```diff
@@ -22,6 +25,25 @@ void AP_BattMonitor_SMBus::read(void)
...
+    }
+}
+
+// reads the pack full charge capacity
+// returns true if the read was successful, or if we already knew the pack capacity
+bool AP_BattMonitor_SMBus::read_full_charge_capacity(void)
+{
+    uint16_t data;
+
+    if (_full_charge_capacity) {
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: Nitpicking: same 0 compare

**Comment by Pedals2Paddles** [comment]
@OXINARF The existing behavior of the stock Solo is to always use the SMBUS reported battery capacity.  Anything the user entered into the BATT_CAPACITY parameter would be overwritten by the SMBUS reported value.  If no SMBUS capacity data is available, it would remains whatever the parameter is set for.  This is probably the desired behavior for all the the far edge use cases. The majority of Solo users don't even know there is a battery capacity parameter to change.

That said, I can see where it would be useful. Perhaps this can go through in it's current implementation to fix the defect.  Then further enhancement can come down the road?

**Comment by WickedShell** [comment]
@OXINARF I'll address the rest in a bit, but as far as the parameters, we are expecting to get them from the SMBus device. The reason I wanted to reset it, is that on a SMBus battery we should always be reading those fields from the battery.  If we only reset the parameter after we got it, there isn't anyway for a user to realize they aren't downloading the correct parameters. IE we've had a bug in the solo driver where it wasn't reseting the value, so all the solo on master users were stuck with just seeing 5200, even though the batteries were actually, this wasn't noticed for quite a long time because there was a value there. I think if you are using a battery system that is meant to provide the information, then we should be making it really clear if it's working.

**Comment by WickedShell** [comment]
@Pedals2Paddles has tested the most recent change set, and it is working as expected to on a Solo.

**Comment by OXINARF** [comment]
@WickedShell @Pedals2Paddles Ok, let me try to explain the behaviors I'm envisioning (and that is almost there in this PR):

- system boots for the first time with default values
- user doesn't set none of this two parameters
- SMBus drivers reads them from battery and sets the parameter value (but doesn't save it! I think you are missing this important bit)
- reboot the system and this entire cycle repeats

- system boots with default parameters
- user sets (and saves!) one of these parameters
- SMBus driver reads the values from battery but doesn't set the parameters because they aren't default value anymore
- reboot the system and it will load the values the user setted. If changed to default, driver will set them to the value read from battery

Does this make sense? For full battery capacity it may not make much sense, but for serial number I think it does. Most people that care about that already put stickers on batteries to identify them with a number (1, 2, 3, etc.), so having the same number on ArduPilot would be helpful (instead of having to look at where the serial number is in the battery). Still, it allows anyone to use the serial number if that is their preference - and you can change between the two, it is up to you!

**Comment by WickedShell** [comment]
@OXINARF we explicitly don't want to save the numbers actually!

So let's treat them seprately for a moment.

Serial Number: When you swap batteries in the field to fly a different battery we need to know what serial number it is. The point of a smart battery is it self identifies, and will track. We aren't building an internal mapping that battery id 15923 is actually user battery 5. The user has to label the batteries to match. If the parameter was user set/maintained then we gain no value from having it, as users will forget to do it. For the record I made the change to the driver such that serial number is always reset if when we request it it doesn't match the expected number.

Full Charge Capacity: This number actually changes over time, it's not always a 5200 value. Instead this is an estimate on how much charge the battery is currently capable of holding. For example one of Matt's batteries has degraded to only being 4008 mAh according to the BMS system. Again this will change every time you swap batteries out, and over the lifetime of the aircraft. We want to always fetch the most recent battery. If you are using a smart battery system you shouldn't be trying to override the capacity estimates. That absolutely shoots the smarts, and makes it pointless :)

Does that make sense? Happy to discuss further here, or on the dev call if needed. (I'll just tag this for the call now, and if we've resolved it before then we can skip it :) )

**Comment by WickedShell** [comment]
@OXINARF Also worth noting that Solo batteries don't map the internal serial number to the one printed on the package in any meaningful way :(

**Comment by OXINARF** [comment]
@WickedShell I wasn't able to be in the dev call yesterday so I don't know what was concluded, but I'll answer your post.

> we explicitly don't want to save the numbers actually!

If they come from the driver I agree completely. If a user sets them, why not?

> We aren't building an internal mapping that battery id 15923 is actually user battery 5. The user has to label the batteries to match.
> Also worth noting that Solo batteries don't map the internal serial number to the one printed on the package in any meaningful way :(

You just gave me my best argument. You grab a log, it has the serial number of the Solo battery, how do you know what battery it belongs to if it's not printed anywhere? If a user has to write down the match between a serial number and a user id, why not allow him to set his user id directly in ArduPilot?

>  If the parameter was user set/maintained then we gain no value from having it, as users will forget to do it.

If a user forgets to change the parameter that is his own fault. He can just change the parameter back to default to start using the serial number from the driver.

I don't understand why you don't want flexibility, what is the problem of allowing two methods of operation?

Regarding full capacity I'm not going to argue much about it because I don't have a good case. Although I'll say that BMS are frequently wrong in their reported values. And so, again, I don't see the issue with having flexibility and allowing two ways of working.

**Comment by Pedals2Paddles** [comment]
@OXINARF users that currently keep a log of their batteries are doing so manually. They put a sticker on their battery with a made up ID# or use a paint marker.  Then with each flight, they have to keep notes of which battery they used. Then update their logs by hand.  Pulling the SN from the battery eliminates all of those steps.  The SN is automatically read from the battery and placed in the parameter, which gets it in the dataflash log.  The dataflash log is automatically read, and automatically populates the user's battery log. The user would put the SN on their battery sticker rather than an arbitrary 1,2,3,4,5,etc so it matches up with reality.  There are no circumstances I can think of where a user should, or would want to override the SMBUS reported serial number.  It needs to auto fill if it's there maintain integrity of the information.

However, I think there is probably a use case for manually entering the battery SN **where there is no smart battery to update the parameter automatically**.  That would allow all other ArduPilot users to start doing this automatic battery logging if they want to. The user would need to use the GCS before flight to manually key in whatever arbitrary ID# they have on their battery.  They could then use the dataflash log to update their battery logs.  For this to work, I think we just need to make sure that the Battery SN is not being overwritten to zero or -1 in the absence of a smart battery.


Regarding the Battery Capacity, I believe it is doing exactly what it should be doing IMO. Read the capacity reported by the smart battery, and set the battery capacity parameter to match.  Requiring the user to manually enter the capacity, would completely defeat one of the most critical functions of the smart battery.  In all the batteries I've burned through on the Solo, I've never seen the reported capacity be wildly off base.  I test them all time by running them down to see if 0% is legitimately where it should be.  And it's always pretty darn good.  I can say with 100% certainty that if you just arbitrarily leave a fixed number in there, it will absolutely be wrong, and vehicle will crash from fuel starvation.

If the BATT_MONITOR parameter is set for one of the smart battery modes, @WickedShell has written it to fail safely if no battery capacity is read from the smart battery.  So if the battery is malfunctioning or someone has hot wired a Hobby King battery into it, and no capacity information is received, it sets the BATT_CAPACITY parameter for zero.  This would immediately clue the user in that something is wrong with the battery.  It will prevent them from flying off with an arbitrary and incorrect default capacity.  The user would need to take positive action to override this and take off. 

There again are no circumstances I can think of where a user should/would want to manually override the reported capacity with a manually entered value. But if you feel really strongly that we should let users do it anyway, I suppose maybe we can just read and update the battery capacity once on startup rather than in the loop continuously.  That would let it do it's thing, but still let the user override it if they don't like it.

**Comment by OXINARF** [comment]
@Pedals2Paddles Sorry but your comment is a contradiction in itself. You (and @WickedShell) are saying that the way you envision the user modus operandi is to change no parameters. But then you say that code needs to do a reset of the parameter to default because user changed it. Which is it? Does the user change parameters or not?

> Read the capacity reported by the smart battery, and set the battery capacity parameter to match. Requiring the user to manually enter the capacity

This tells me you aren't understanding what I am proposing at all. Why would the user be **required** to enter the capacity?

**Comment by Pedals2Paddles** [comment]
@OXINARF It isn't changing the parameter to default to get around user changes. I think that's why we're debating this from different directions. It's just comparing the current parameter with the value reported by the smart battery.  If different, it updates the parameter with the smart battery reported value. No user input is needed or expected. And anything the user enters will be overwritten the next time the value is read and compared in the loop.  if I open up MP and change the SN or capacity parameter, write the change, then refresh the parameters 10 seconds later, the value will have already been overwritten back to the smart battery reported values.  This is the current expected behavior of the stock solo too.  It has always done this.  The SN would be something new not found in a stock Solo.

If we want the user to be able to override the smart battery reported SN or capacity, the best thing to do is probably just have it read and update the parameter once on startup, rather and constantly in the loop.  This might save a few clock cycles too.  In this configuration, the vehicle will always get the new values on startup.  But the user can change them at will without them being immediately overwritten.

**Comment by WickedShell** [comment]
@OXINARF we only reset the serial number on smart batteries, normal batteries support exactly the interface you want.

The reason to reset, is so that its clear if the smart battery has failed to communicate those values back to the autopilot. Without doing that reset I fail to see how I have any indication if the interface failed or not. This is exactly how we had bad battery capacity on master for solo for quite awhile, no one noticed because the parameter was roughly reasonable, but it wasn't updating.

Regarding serial numbers and battery labeling. Please remember that the solo battery numbers might not map out correctly, but that doesn't mean a battery from a different manufacturer won't perfectly map to the number printed on the battery. All I'm proposing is that a smart battery user use the serial number that's in the smart section. I don't see anyway to allow you to override this that doesn't break automatically fetching it. What confuses me is I can't come up with any reasonable use case to allow the user to specify this on a smart battery. As far as I can tell it would be like asking a user to set the COMPASS_DEVID parameter. We determined what it was by talking to the device, there isn't a question what the ID should be, we need to report that value.

**Comment by OXINARF** [comment]
I talked with @WickedShell on Gitter and finally we reached a conclusion: my proposal can only break your method of operation if an user loads an entire parameter file (which is a dumb thing to do).

I think that protecting against that is an overprotection we don't do anywhere else (you can run multiple EKFs, you can load completely invalid calibration values, etc.) but I'm tired of fighting over this so I'll let it go. I'll review it later today or tomorrow and merge it in.

**Comment by Pedals2Paddles** [comment]
LOL I agree, if someone blindly loads a parameter file, the battery serial number will be the least of their problems anyway.   I'm not following what you're proposing. But need to get back to work too. I trust you guys so LGTM!

**Comment by rmackay9** [comment]
I don't have strong feelings on this but I think:

- for the original Solo battery monitor I allowed the user to override the capacity and we would use that number from then on unless the user set the capacity back to zero (I think).  I was worried that the reported capacity might be inaccurate so I wanted to allow an override.  @Pedals2Paddles says it always seems correct so removing this ability is probably OK.
- I worry that if the capacity read fails and we set the battery capacity to zero and the user has no way override the battery capacity that they can no longer use the battery capacity based battery failsafe.
- I don't think we should let the user override the battery serial number if we can successfully read it from the battery.  I.e. for the serial number - the value from the battery should be used over anything the user inputs.

**Comment by WickedShell** [comment]
@rmackay9 that's actually why it's been broken on master for solo... Everyone's been passing around param files which have the capacity set, which they then loaded which forced the parameter. I think this is honestly not behavior we should be accepting from smart batteries.

My current take on this is that if the capcity isn't available then we have massive problems. My 2 proposals on this are A) I can modify the code so you can set a parameter if we haven't managed to query the battery yet. B) I'm planning on adding some SMBus arming checks that allow us to have established that we can actually read current/capacity/voltages at a minimum before arming. I was planning the later as another/later PR though. (Once I added fetching battery status fields so we could get the health bits out of that as well).

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 7]
```diff
@@ -2,7 +2,49 @@
...
+void AP_BattMonitor_SMBus::read(void)
+{
+    // nothing to be done here for actually interacting with the battery
+    // however we can use this to set any parameters that need to be set
+
+    if (_serial_number != _mon._serial_numbers[_state.instance]) {
+        _mon._serial_numbers[_state.instance].set_and_notify(_serial_number);
+    }
+
+    if (_full_charge_capacity != _mon._pack_capacity[_state.instance]) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: I can make this be ```if (_full_charge_capacity != 0 && (_full_charge_capacity != _mon._pack_capacity[_state.instance])) {``` if that makes everyone happier.

**Comment by rmackay9** [comment]
I've added a new To-Do item for MP to display the addition cell voltages and battery temperature.
https://github.com/ArduPilot/MissionPlanner/issues/1530

**Comment by rmackay9** [comment]
I did some additional testing with my Maxell batteries (I have two) and found the serial number was updating correctly (and it's different for each battery).  Capacity and individual cell voltages were also present.

smartbatt1
smartbatt2

**Comment by Pedals2Paddles** [comment]
When I get home later, I will do some tests on a stock solo to see how it handles no response from the smart battery today.  I know the % remaining reports  -1.

**Comment by OXINARF** [comment]
> I'm not following what you're proposing.

@Pedals2Paddles Maybe you are missing the same thing as @WickedShell did: the values read from the battery aren't saved in the parameters, every time you reboot the parameter is back to default - unless a user explicitly writes it.

> I don't think we should let the user override the battery serial number if we can successfully read it from the battery. I.e. for the serial number - the value from the battery should be used over anything the user inputs.

@rmackay9 Why? The serial number is only used for logging, what's the harm in allowing a user to set their own value (even if the battery has one)?

@WickedShell Your proposal isn't enough because you are setting capacity to 0 on construction. But your idea is what I've said since the beginning for both parameters... I agree with having more battery pre-arm checks though.

Like I said, I'm willing to let it to, but I'll say that I have a hard time understanding why we would allow the user to override the capacity but not the serial number - one has influence on ArduPilot operation, the other doesn't, but we restrict the one that has no influence? It would be like restricting the log mask because you might disable important logging bits, but allowing changing all other flight parameters.

**Comment by WickedShell** [comment]
@OXINARF I'm back at the place of if the battery fails to respond I want to know, and param file loading will accidentally trash the value... My proposal was just meant to allow you to override the param value on a particular flight.

**Comment by OXINARF** [comment]
@WickedShell That would be better than never allowing the user to set it, but doesn't fix @rmackay9's concern.

**Comment by Pedals2Paddles** [comment]
OK, I'm definitely not seeing where/how the parameter is changing back to default on reboot. You guys know more about what this code is doing than me, so I defer to you.  My presumption was that it writes the parameter, period, done.  If the parameter is changing itself back to default on reboot, I don't think that is a good behavior. 

Ultimately, my vision for this (maybe in the future if not now) is as follows:

* If BATT_MONITOR set for a smart battery, read capacity, state of charge (%), and serial # on startup (maybe loop a few times if needed to get it). No need to continuously read capacity and SN once received at startup. This will save some clock cycles and allow operator override.
* If values received from the smart battery, write them to the parameters and save.  Mavlink message GCS notifying operator of received capacity, volts, %, and SN.
* If any values not received from smart battery, write 0 to the parameter and save.  Mavlink message GCS notifying operator of problem.  Operator will be unable to fly without addressing the problem. This is the "fail safest" approach in my opinion. But is it too restrictive? Thoughts?
* _Nothing should be arbitrarily overwritten with default values on boot_. Those defaults have no application to anything and can totally mislead the operator into a crash.  The operator should have to enter something or disable the failsafe in the even the smart battery has failed. 
* With capacity and SN only being checked and written to the parameter once on boot, this allows the operator to key in their own values to override if they so choose. For example, if they know they capacity of the battery, they can key in the value to get around an SMBUS malfunction.  Or if they've hooked up a Hobby King battery for testing, or or or etc.

So basically, this approach accomplishes the expected automation with now operator input required.  It accomplishes notifying the operator if something didn't work as expected.  And it allows the operator to override it.

**Comment by OXINARF** [comment]
@Pedals2Paddles I think you are over-engineering it :slightly_smiling_face: 

> If any values not received from smart battery, write 0 to the parameter and save. 

How long do you wait? How many tries? What happens if the user tries to arm? The parameter would have a value that isn't default.
I think the current way of setting but not saving the values to parameters is perfect, I just think it should allow users to set, save and use them.

**Comment by rmackay9** [comment]
Maybe we should just merge this.  It adds a lot of value and none of this is written in stone.  Let's merge and see if users complain or ask for changes.

**Comment by Pedals2Paddles** [comment]
LOL good idea @rmackay9.  As it is now, it works well and is needed. But I think this discussion shows the whole concept of smart batteries has lots of potential and a lot of variability. It is clearly a way bigger discussion than comments on a PR.  We should probably work through the scenarios and combine all our  ideas over time on gitter and discuss.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp [hunk 8]
```diff
@@ -18,6 +38,22 @@ bool AP_BattMonitor_SMBus::read_temp(void)
     return false;
 }
 
+// reads the serial number if it's not already known
+// returns true if the read was successful or the number was already known
+bool AP_BattMonitor_SMBus::read_serial_number(void) {
+    uint16_t data;
+
+    // don't recheck the serial number if we already have it
+    if (_serial_number != 0) {
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: Isn't 0 a valid serial number? I thought before that 0 for the parameter default would be better than -1, but then I thought that the battery might have the serial number 0.

Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp]: I think it is. I actually had the thought while working on it, then forgot to correct it. Will address that real quick.

**Comment by WickedShell** [comment]
@OXINARF I addressed the serial_number being 0 comment in it's own commit. It can be squashed, but just kept it out to make it easy to review I left it as it's own.

**Comment by OXINARF** [comment]
@WickedShell I've changed the `_serial_number` initialization to the class declaration instead of the constructor, squashed and merged. I hope you don't mind.

Thanks!

**Comment by Pedals2Paddles** [comment]
Thanks @OXINARF and @WickedShell
